### PR TITLE
Fix 422

### LINKED ISSUE #422: Material gets duplicated when dragging it just a bit on the Construction tab
Issue overview
--------------

Reported by @eringold on #415 

Select a material already in a construction and drag to layer drop zone -> it will be copied. After, select a material and drag _just a tiny bit_, and it will be added again. 

## Current Behavior

1 2 1rc_cons
3. #420

## Steps to Reproduce

See gif above

### Environment

Some additional details about your environment for this issue (if relevant):
 - Platform (Operating system, version): Windows at least, not Ubuntu
 - Version of OpenStudioApplication (if using an intermediate build, include SHA): 1.2.1-rc1

### PR DESCRIPTION
**Pull request by macumber**
Fix #422

### CODE REVIEW

File: src/openstudio_lib/OSItem.hpp [hunk 1]
```diff
@@ -59,18 +59,26 @@ class OSItemId
...
+  void setPosition(int position);
+
   bool operator==(const OSItemId& other) const;
 
  private:
   QString m_itemId;
   QString m_sourceId;
   QString m_otherData;
   bool m_isDefaulted;
+  boost::optional<int> m_position;
```
Comment (jmarrec) [inline src/openstudio_lib/OSItem.hpp]: OSItemId now stores a position. OSItem forwards to its m_itemId. :+1:

File: src/openstudio_lib/OSItem.cpp [hunk 2]
```diff
@@ -421,12 +437,12 @@ void OSItem::setOSItemType(OSItemType osItemType) {
...
 
-int OSItem::position() const {
-  return m_position;
+boost::optional<int> OSItem::position() const {
+  return m_itemId.position();
 }
 
 void OSItem::setPosition(int position) {
-  m_position = position;
+  m_itemId.setPosition(position);
```
Comment (jmarrec) [inline src/openstudio_lib/OSItem.cpp:452]: Forwards

File: src/openstudio_lib/OSItem.cpp [hunk 3]
```diff
@@ -102,7 +109,8 @@ QString OSItemId::otherData() const {
 
 QString OSItemId::mimeDataText() const {
   QString isDefaultedString((m_isDefaulted ? "True" : "False"));
-  QString result = m_itemId + "," + m_sourceId + "," + isDefaultedString + "," + m_otherData;
+  QString positonString((m_position ? QString::number(*m_position) : "None"));
+  QString result = m_itemId + "," + m_sourceId + "," + isDefaultedString + "," + positonString + "," + m_otherData;
```
Comment (jmarrec) [inline src/openstudio_lib/OSItem.cpp]: Serialize new field to mimeDataText

File: src/openstudio_lib/OSItem.cpp [hunk 4]
```diff
@@ -80,7 +80,14 @@ OSItemId::OSItemId(const QMimeData* mimeData) : m_isDefaulted(false) {
...
   }
-  for (int i = 3; i < strings.size(); ++i) {
+  if (strings.size() > 3) {
+    if (strings[3] == "None") {
+      m_position.reset();
+    } else {
+      m_position = strings[3].toInt();
+    }
+  }
+  for (int i = 4; i < strings.size(); ++i) {
```
Comment (jmarrec) [inline src/openstudio_lib/OSItem.cpp]: Read back new field from mimeData.

File: src/openstudio_lib/ConstructionObjectVectorController.cpp [hunk 5]
```diff
@@ -149,75 +151,61 @@ void ConstructionObjectVectorController::insert(const OSItemId& itemId, int posi
...
-      model::LayeredConstruction construction = m_modelObject->cast<model::LayeredConstruction>();
-      std::vector<model::Material> layers = construction.layers();
-      auto it = std::unique(layers.begin(), layers.end());
-      deleteExisting = (it == layers.end());
-      if (!deleteExisting) {
-        LOG(Debug, "Layers aren't unique, so we can't remove the existing one, occured for " << m_modelObject->nameString());
-      } else {
-        LOG(Debug, "Layers are unique, occured for " << m_modelObject->nameString());
-      }
+      fromPosition.reset();
```
Comment (jmarrec) [inline src/openstudio_lib/ConstructionObjectVectorController.cpp]: (I guess that makes it clearer, but wouldn't it be boost::none to begin with?)

File: src/openstudio_lib/ConstructionObjectVectorController.cpp [hunk 6]
```diff
@@ -149,75 +151,61 @@ void ConstructionObjectVectorController::insert(const OSItemId& itemId, int posi
...
-    construction.insertLayer(position, *material);
   }
 }
 
 void ConstructionObjectVectorController::onReplaceItem(OSItem* currentItem, const OSItemId& replacementItemId) {
   if (m_modelObject) {
 
+    // currentItem position is set by OSDropZone::setItemIds
+    boost::optional<int> toPosition = currentItem->position();
+    boost::optional<int> fromPosition = replacementItemId.position();
```
Comment (jmarrec) [inline src/openstudio_lib/ConstructionObjectVectorController.cpp]:
```suggestion
    boost::optional<int> toPosition_ = currentItem->position();
    boost::optional<int> fromPosition_ = replacementItemId.position();
```

**Review by jmarrec — APPROVED** [review]

File: src/openstudio_lib/ConstructionObjectVectorController.cpp (hunk 6 above)
Comment (macumber) [reply src/openstudio_lib/ConstructionObjectVectorController.cpp]: Is this a standard you want to follow?  Trailing `_` on optionals? Or is that to indicate it is a local variable?

Comment (jmarrec) [reply src/openstudio_lib/ConstructionObjectVectorController.cpp]: I used _optional before which i liked better. But apparently anything starting with a _ is reserved for standard library implementers. So optional_ is what I use now. 
I find it easier to understand what is an optional that way

**Pull request merged**